### ISSUE TITLE
spark_connect failed on OS X:Failed to launch Spark shell. Ports file does not exist.

### ISSUE THREAD

**Issue by feng5** [rstudio/sparklyr#114]
```
> library(sparklyr)
> library(dplyr)

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

> sc <- spark_connect(master = "local")
Error in sparkapi::start_shell(master = master, spark_home = spark_home,  : 
  Failed to launch Spark shell. Ports file does not exist.
    Path: /Users/13k/Library/Caches/spark/spark-2.0.0-preview-bin-hadoop2.6/bin/spark-submit
    Parameters: --jars, '/Users/13k/R/sparklyr/java/rspark_utils.jar', --packages, 'com.databricks:spark-csv_2.11:1.3.0','com.amazonaws:aws-java-sdk-pom:1.10.34', sparkr-shell, /var/folders/sd/2_0mf8p100v7dl3p_2cs5m_80000gn/T//RtmpXUplRl/file1323016779f8e.out
    
Ivy Default Cache set to: /Users/13k/.ivy2/cache
The jars for the packages stored in: /Users/13k/.ivy2/jars
:: loading settings :: url = jar:file:/Users/13k/Library/Caches/spark/spark-2.0.0-preview-bin-hadoop2.6/jars/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
com.databricks#spark-csv_2.11 added as a dependency
com.amazonaws#aws-java-sdk-pom added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent;1.0
	confs: [default]
Exception in thread "main" java.io.FileNotFoundException: /Users/13k/.ivy2/cache/resolved-org.apache.spar
```
sessioninfo:
```
> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.11.4 (El Capitan)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.5.0     sparklyr_0.2.24

loaded via a namespace (and not attached):
 [1] magrittr_1.5    R6_2.1.2        assertthat_0.1  rsconnect_0.4.3
 [5] parallel_3.3.1  DBI_0.4-2       tools_3.3.1     withr_1.0.2    
 [9] yaml_2.1.13     rappdirs_0.3.1  tibble_1.1      Rcpp_0.12.5    
[13] config_0.1.0    sparkapi_0.3.12
```

**Comment by funnng** [comment]
and i tested sparklyr in terminal with `sudo R`, and it works well.
it's a permission issue about RStudio or something else?

**Comment by javierluraschi** [comment]
The error makes me think that this could be related with loading spark packages; therefore, I would try connecting without any additional packages as follows:

```
sc <- spark_connect(master = "local", config = list())
```

Would this work for you?

**Comment by funnng** [comment]
@javierluraschi yea it can connect by using your code but when i try `copy_to` it said `Error: could not find function "copy_to"` and load dplyr package then copy_to, pop another error `Error: java.lang.ClassNotFoundException: Failed to find data source: com.databricks.spark.csv`

**Comment by javierluraschi** [comment]
@funnng this scopes the problem to loading spark packages. The most common cause would be that something is blocking downloads for that user.

That said, a fix included in `0.2.28` that will allow you to perform `copy_to` even when packages are not loaded. However, not loading packages will decrease `copy_to` performance significantly, but we are also working on improving this in the near future to avoid requiring additional packages.

**Comment by desertnaut** [comment]
I had a similar issue while using the CDH-provided Spark, resolved when I switched to a Databricks prebuilt distro. I got some interesting error messages when trying the arfument ``config==list()``, as suggested above. You may want to have a look at the full details in [this SO thread](http://stackoverflow.com/questions/38102921/can-sparklyr-be-used-with-spark-deployed-on-yarn-managed-hadoop-cluster/) (used version ``0.2.28``).

**Comment by javierluraschi** [comment]
@desertnaut @funnng `sparklyr` version `0.2.29` adds support for CDH. CDH is currently under testing on our end, but if you are interested, please feel free to go ahead and give this a try.

**Comment by desertnaut** [comment]
Thanx - did ``0.2.29`` just come out? I installed it yesterday, and I got ``0.2.28``.

**Comment by javierluraschi** [comment]
@desertnaut Yes, just bumped the version with this commit https://github.com/rstudio/sparklyr/commit/1726b24466ce1cc9e296bebb80d2b7de12f1f966. Although the relevant commit is this one https://github.com/rstudio/sparklyr/commit/036101f35d1757bd53f4f241e12973fd53b28de0 but it's hard to tell if you have the changes without a version bump.

**Comment by desertnaut** [comment]
OK, I'll take the new version for a test drive with CDH asap. BTW, thanks for this great work...!

**Comment by javierluraschi** [comment]
@desertnaut there were a couple additional issues that affected the `2.0.0-preview` version. If you were using this version, then you would want to upgrade `sparklyr`; otherwise, you can ignore this comment.

**Comment by mikejacktzen** [comment]
I'm also running into this issue
```

devtools::install_github("rstudio/sparklyr",dependencies =TRUE)

library(sparklyr)
spark_install(version = "1.6.2")

# Installation complete.
# Warning message:
#   In value[[3L]](cond) : Failed to set logging settings

library(sparklyr)
sc <- spark_connect(master = "local",config = list())
# Error in sparkapi::start_shell(master = master, spark_home = spark_home,  : 
#                                  Failed to launch Spark shell. Ports file does not exist.

# iris_tbl <- copy_to(sc, iris)
```

```
R version 3.2.3 (2015-12-10)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: CentOS release 6.7 (Final)

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] sparklyr_0.2.31

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.6     sparkapi_0.3.22 digest_0.6.9    dplyr_0.5.0     withr_1.0.2    
 [6] rprojroot_1.0-2 assertthat_0.1  rappdirs_0.3.1  R6_2.1.2        DBI_0.4-1      
[11] git2r_0.14.0    magrittr_1.5    httr_1.2.1      curl_1.1        config_0.1.0   
[16] devtools_1.10.0 tools_3.2.3     readr_0.2.2     parallel_3.2.3  yaml_2.1.13    
[21] memoise_1.0.0   tibble_1.1     
```

**Comment by koaning** [comment]
On mac I am experiencing the same issue but only when I edit the `config.yml` file. In particular, it is the final line here that is throwing the error:

```
default:
  sparklyr.cores.local: 8
  sparklyr.shell.driver-memory: 4GB
```

**Comment by selva86** [comment]
I am getting this error on Mac.. What seems to be the problem?
``` r
library(sparklyr)
library(dplyr)
spark_install_dir()
[1] "/Users/selvaprabhakaran/Library/Caches/spark"
Sys.setenv(SPARK_HOME="/Users/selvaprabhakaran/Library/Caches/spark")

sc <- spark_connect(master = "local", config = list(), version = "1.6.1")
Error in start_shell(master = master, spark_home = spark_home, spark_version = version,  : 
  Failed to launch Spark shell. Ports file does not exist.
    Path: /Users/selvaprabhakaran/Library/Caches/spark/bin/spark-submit
    Parameters: --class, sparklyr.Backend, '/Library/Frameworks/R.framework/Versions/3.3/Resources/library/sparklyr/java/sparklyr-1.6-2.10.jar', /var/folders/q5/gkgbtsnx0mjg2rq_0bdwct4c0000gp/T//RtmpgUvlqC/file127764421f8b1.out
    
sh: /Users/selvaprabhakaran/Library/Caches/spark/bin/spark-submit: No such file or directory
In addition: Warning message:
In normalizePath(file.path(spark_home, "bin", spark_submit)) :
  path[1]="/Users/selvaprabhakaran/Library/Caches/spark/bin/spark-submit": No such file or directory
```

sessioninfo:
``` r
sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.10.5 (Yosemite)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] sparklyr_0.3.3

loaded via a namespace (and not attached):
 [1] readr_1.0.0     magrittr_1.5    R6_2.1.2        assertthat_0.1 
 [5] rprojroot_1.0-2 parallel_3.3.1  DBI_0.4-1       tools_3.3.1    
 [9] withr_1.0.2     dplyr_0.5.0     rappdirs_0.3.1  yaml_2.1.13    
[13] tibble_1.1      Rcpp_0.12.6     config_0.2      digest_0.6.10  
```

**Comment by selva86** [comment]
Turns out there was 2 versions of spark in my SPARK_HOME="/Users/selvaprabhakaran/Library/Caches/spark"

So, I set it to: 
``` r
Sys.setenv(SPARK_HOME="/Users/selvaprabhakaran/Library/Caches/spark/spark-1.6.1-bin-hadoop2.6")
```
But, now I get different exceptions.. 
``` r
> sc <- spark_connect(master = "local", config = list(), version = "1.6.1")
Error in start_shell(master = master, spark_home = spark_home, spark_version = version,  : 
  Failed to launch Spark shell. Ports file does not exist.
    Path: /Users/selvaprabhakaran/Library/Caches/spark/spark-1.6.1-bin-hadoop2.6/bin/spark-submit
    Parameters: --class, sparklyr.Backend, '/Library/Frameworks/R.framework/Versions/3.3/Resources/library/sparklyr/java/sparklyr-1.6-2.10.jar', /var/folders/q5/gkgbtsnx0mjg2rq_0bdwct4c0000gp/T//RtmpgUvlqC/file127767a2bc95b.out
    
Exception in thread "main" java.lang.NoClassDefFoundError: org/apache/spark/launcher/Main
Caused by: java.lang.ClassNotFoundException: org.apache.spark.launcher.Main
	at java.net.URLClassLoader$1.run(URLClassLoader.java:202)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:190)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:306)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:301)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:247)
```

**Comment by selva86** [comment]
Resolved it by re-installing JDK 8 from this [link](http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html) looks like there was a conflict with the Java version used by terminal and by setting `Sys.setenv(SPARK_LOCAL_IP="127.0.0.1")`

**Comment by mkaja** [comment]
hi,
 i am trying to connect to emr master server running yarn and getting this error -- any help?

> sc <- spark_connect(master = "yarn-client")
Error: Failed during initialize_connection() java.net.ConnectException: Call From 98bda6519109/127.0.0.1 to 0.0.0.0:8032 failed on connection exception: java.net.ConnectException: Connection refused; For more details see:  http://wiki.apache.org/hadoop/ConnectionRefused
	at sun.reflect.GeneratedConstructorAccessor11.newInstance(Unknown Source)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.apache.hadoop.net.NetUtils.wrapWithMessage(NetUtils.java:791)
	at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:731)
	at org.apache.hadoop.ipc.Client.call(Client.java:1472)
	at org.apache.hadoop.ipc.Client.call(Client.java:1399)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:232)
	at com.sun.proxy.$Proxy10.getClusterMetrics(Unknown Source)
	at org.apache.hadoop.yarn.api.impl.pb.client.ApplicationClientProtocolPBClientImp

**Comment by javierluraschi** [comment]
@mkaja looks like there is no access from the machine where the r-session is running to the emr driver node. If you are running in emr, I would recommend you install rstudio in emr. For instance, see: https://blogs.aws.amazon.com/bigdata/post/Tx37RSKRFDQNTSL/Statistical-Analysis-with-Open-Source-R-and-RStudio-on-Amazon-EMR

**Comment by JackHo327** [comment]
Hi @selva86, could you tell me which version of JAVA did you re-installed, because I also have the same problem that you met before.
I am on windows 10
my java  version is 
> java version "1.8.0_101"
Java(TM) SE Runtime Environment (build 1.8.0_101-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.101-b13, mixed mode)

**Comment by selva86** [comment]
@JackHo327 I installed 'jdk-8u101-macosx-x64.dmg' from [link](http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html). You might've to install 'jdk-8u101-windows-x64.exe' if your PC is 64bit.

**Comment by JackHo327** [comment]
Thank you @selva86 .
but unfortunately, still got that problem.
so strange because I can run this at spark 1.6.x

**Comment by javierluraschi** [comment]
@JackHo327 how are you setting the `SPARK_HOME`? sparklyr looks at the system environment variable `SPARK_HOME` so if this happens to point to the `1.6.*` version, it would also use the wrong jars while deploying to Spark.

**Comment by mkaja** [comment]
*Sys.setenv(HADOOP_CONF_DIR= "/etc/hadoop/conf")*


*Sys.setenv(SPARK_HOME=
"/home/user-id/.cache/spark/spark-1.6.0-bin-hadoop2.6")*

**Comment by shuaiwang88** [comment]
@selva86 I have two installed on Mac as well. Solved by deleting the brew installed one and restart the Mac.

**Comment by javierluraschi** [comment]
The connection code has been significantly simplified starting with `sparklyr` version `0.4.9`.

Please upgrade `sparklyr` using `devtools::install_github("rstudio/sparklyr")` and report back if any connection issues arise.

**Comment by JackHo327** [comment]
@javierluraschi  my problem solved wi 0.4.9 version.

ps. need to point out the **spark.sql.warehouse.dir**

**Comment by Jim89** [comment]
Just a note on this from me. Using the CRAN version (0.4 at time of writing) and specifying the `config = list()` option as advised above was not loading the necessary `csv` [spark packages](https://spark-packages.org/package/databricks/spark-csv) and hence it was not possible for me to read a CSV from hdfs into a spark dataframe with `spark_read_csv`. 

Installing the dev. version from Github solved this for me. I'm on Spark 1.6.0.
